datalad_fuse/fuse_.py

```diff
@@ -126,34 +126,48 @@ def getattr(self, path, fh=None):
             else:
                 iadok = is_annex_dir_or_key(path)
                 if iadok is not None:
-                    topdir, dir_or_key = iadok
-                    if dir_or_key == "key":
-                        # needs to be open but it is a key. We will let fsspec
-                        # to handle it
-                        pass
-                    elif dir_or_key == "dir":
+                    if isinstance(iadok, AnnexKey):
+                        if iadok.size is not None:
+                            lgr.debug("Got size from key")
+                            r = mkstat(
+                                is_file=True,
+                                size=iadok.size,
+                                timestamp=self._adapter.get_commit_datetime(path),
+                            )
+                        else:
+                            # needs to be open but it is a key. We will let
+                            # fsspec handle it
+                            pass
+                    elif isinstance(iadok, AnnexDir):
                         # just return that one of the top directory
                         # TODO: cache this since would be a frequent operation
-                        r = self._filter_stat(os.stat(topdir))
+                        r = self._filter_stat(os.stat(iadok.topdir))
                     else:
-                        raise AssertionError(f"Unexpected dir_or_key: {dir_or_key!r}")
+                        raise AssertionError(f"Unexpected iadok: {iadok!r}")
                 elif self.is_under_git(path):
                     lgr.debug("Path under .git does not exist; raising ENOENT")
                     raise FuseOSError(ENOENT)
         if r is None:
+            fsspec_file = None
             if fh and fh >= self._counter_offset:
                 lgr.debug("File already open")
                 fsspec_file = self._fhdict[fh]
                 to_close = False
             else:
-                # TODO: it is expensive to open each file just for `getattr`!
-                # We should just fabricate stats from the key here or not even
-                # bother???!
-                lgr.debug("File not already open")
-                with self.rwlock:
-                    fsspec_file = self._adapter.open(path)
-                to_close = True
-            if fsspec_file:
+                _, key = self._adapter.get_file_state(path)
+                if key.size is not None:
+                    lgr.debug("Got size from key")
+                    r = mkstat(
+                        is_file=True,
+                        size=key.size,
+                        timestamp=self._adapter.get_commit_datetime(path),
```

Comment:
filed a dedicated https://github.com/datalad/datalad-fuse/issues/88 to possibly gain us some performance for healthcheck